diff --git a/simple_sitemap.install b/simple_sitemap.install
index 8f701e1..9c6f8c4 100644
--- a/simple_sitemap.install
+++ b/simple_sitemap.install
@@ -704,3 +704,17 @@ function simple_sitemap_update_8301() {
     ]
   );
 }
+
+/**
+ * Remove batch_process_limit from settings setting if it still exists.
+ */
+function simple_sitemap_update_8302() {
+  $settings = \Drupal::service('config.factory')->getEditable('simple_sitemap.settings');
+
+  if($settings->get('batch_process_limit') === NULL){
+    return;
+  }
+
+  $settings->clear('batch_process_limit');
+  $settings->save();
+}
